src/backend/executor/nodeShareInputScan.c

```diff
@@ -627,38 +627,50 @@ static void create_tmp_fifo(const char *fifoname)
 /* 
  * As all other read/write in postgres, we may be interrupted so retry is needed.
  */
-static int retry_read(int fd, char *buf, int rsize)
+static int retry_read(int *fd, char *buf, int rsize)
 {
 	int sz;
 	Assert(rsize > 0);
 
 read_retry:
-	sz = read(fd, buf, rsize);
+	sz = read(*fd, buf, rsize);
```

Comment:
Frankly speaking, I'd retry_read() logic simple like this:
    do
    {
        err =read(fd, buf, rsize);
    } while (err == -1 && errno == EINTR);

And leave close() and error handling code in callers of it.

If you insist on this, at least you could modify the function name to reflect the additional close() call and exiting.

I do not why a fd pointer is needed here since elog(ERROR, ...) will quit the progress.

The comment applies to the write change below also.